File: dcget/src/hub/HubConnection.java
package hub;

import java.net.Socket;
import logger.ILogger;
import util.KeyGenerator;

public class HubConnection {

    private IHubEventHandler handler;
    private HubReader reader;
    private HubWriter writer;
    private String nick;

    public HubConnection(IHubEventHandler handler, ILogger logger, String host, int port, String nick) throws Exception {
        this.handler = handler;
        logger.debug("connecting to " + host + ":" + port);
        Socket sock = new Socket(host, port);
        this.reader = new HubReader(sock.getInputStream(), logger);
        this.writer = new HubWriter(sock.getOutputStream(), logger);
        this.nick = nick;
        reader.registerHandler(new LockHandler(this));
        reader.registerHandler(new SRHandler(handler));
        reader.registerHandler(new ConnectToMeHandler(handler));
    }

    public void run() throws Exception {
        reader.read();
    }

    public void onHubConnected(String lock) throws Exception {
        writer.sendKey(KeyGenerator.generateKey(lock.getBytes()));
        writer.sendValidateNick(nick);
        writer.sendVersion("0.01");
        writer.sendMyInfo(nick);
        handler.onHubConnected();
    }

    public void search(String tth) throws Exception {
        writer.sendTTHSearch(nick, tth);
    }

    public void requestPeerConnection(String target) throws Exception {
        writer.sendRevConnectToMe(nick, target);
    }

}

File: dcget++/HubConnection.cpp
#include "HubConnection.h"

#include "LockHandler.h"
#include "SRHandler.h"
#include "ConnectToMeHandler.h"

HubConnection::HubConnection(IHubEventHandler *handler, ILogger *logger, const std::string& host, int port, const std::string& nick)
{
    this->nick = nick;
    this->handler = handler;
    logger->debug("connecting to " + host);
    this->sock = new Socket();
    sock->connect(host, port);
    sock->set_non_blocking(true);
    logger->debug("connected");
    this->writer = std::auto_ptr<HubWriter>(new HubWriter(sock, logger));
    this->reader = std::auto_ptr<HubReader>(new HubReader(sock, logger));
    reader->registerHandler(new LockHandler(this));
    reader->registerHandler(new SRHandler(handler));
    reader->registerHandler(new ConnectToMeHandler(handler));
}

